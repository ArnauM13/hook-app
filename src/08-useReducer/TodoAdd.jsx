import React from 'react'
import { useForm } from '../hooks'

export const TodoAdd = ({onNewTodo}) => {

    // @ts-ignore
    const {description, onInputChange, onResetForm} = useForm({
        description: '' // passem com a form unicament el description i el recuperem
    });

    const onFormSubmit = (event) => { // creem una funció que al fer el submit s'executi
        event.preventDefault(); // fem que no s'envii per default

        if (!description.length) return; // si no tenim re no fem re
        const newTodo = { // creem un nou todo
            id: new Date().getTime(),
            description,
            done: false
        }

        onNewTodo(newTodo); // executem funcio que ve del parent
        onResetForm(); // resetejem l'input del form
    }

    return (
        <form onSubmit={onFormSubmit}>
            <input
                type="text"
                placeholder='¿Que hay que hacer?'
                className='form-control'
                name='description' // afegim el name i el value per poderlos processar
                value={description}
                onChange={onInputChange} // quan faci canvi s'executarà la funció que gestiona des del hook
            />
            <button
                type='submit'
                className='btn btn-outline-primary mt-1'
            >
                Agregar
            </button>
        </form>
    )
}
