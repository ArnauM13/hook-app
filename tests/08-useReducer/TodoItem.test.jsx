import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react';
import { TodoItem } from '../../src/08-useReducer/TodoItem';

describe('Pruebas en el componente <TodoItem />', () => {

    const initialTodo = {
        id: 1,
        description: 'Demo Todo',
        done: false
    };

    const onDeleteTodoMock = jest.fn();
    const onToggleTodoMock = jest.fn();

    beforeEach(() => jest.clearAllMocks());
 
    test('debe de hacer match con el snapshot', () => {

        const {container} = render(
            <TodoItem
                todo={initialTodo}
                onDeleteTodo={onDeleteTodoMock}
                onToggleTodo={onToggleTodoMock}
            />
        );

        expect(container).toMatchSnapshot();
    });

    test('debe de mostrar el Todo pendiente de completar', () => {
        
        render(
            <TodoItem
                todo={initialTodo}
                onDeleteTodo={onDeleteTodoMock}
                onToggleTodo={onToggleTodoMock}
            />
        );

        const liElement = screen.getByRole('listitem');
        expect(liElement.className).toBe('list-group-item d-flex justify-content-between');

        const spanElement = screen.getByLabelText('span');
        expect(spanElement.className).toContain('align-self-center');
        expect(spanElement.className).not.toContain('text-decoration-line-through');
    });

    test('debe de mostrar el Todo completado', () => {
        
        const doneTodo = {
            ...initialTodo,
            done: true
        };

        render(
            <TodoItem
                todo={doneTodo}
                onDeleteTodo={onDeleteTodoMock}
                onToggleTodo={onToggleTodoMock}
            />
        );
        
        const spanElement = screen.getByLabelText('span');
        expect(spanElement.className).toContain('text-decoration-line-through');
    });

    test('el span debe de llamar el ToggleTodo cuando se hace click', () => {
        
        render(
            <TodoItem
                todo={initialTodo}
                onDeleteTodo={onDeleteTodoMock}
                onToggleTodo={onToggleTodoMock}
            />
        );

        const spanElement = screen.getByLabelText('span');
        fireEvent.click(spanElement);
        expect(onToggleTodoMock).toHaveBeenCalledWith(initialTodo.id);

    });

    test('el button debe de llamar el DeleteTodo cuando se hace click', () => {
        
        render(
            <TodoItem
                todo={initialTodo}
                onDeleteTodo={onDeleteTodoMock}
                onToggleTodo={onToggleTodoMock}
            />
        );

        const btnElement = screen.getByLabelText('delete-button');
        fireEvent.click(btnElement);
        expect(onDeleteTodoMock).toHaveBeenCalledWith(initialTodo.id);

    });

});