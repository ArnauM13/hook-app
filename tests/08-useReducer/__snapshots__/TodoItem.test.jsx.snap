// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoItem /> debe de hacer match con el snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
  >
    <span
      aria-label="span"
      class="align-self-center "
    >
      Demo Todo
    </span>
    <button
      aria-label="delete-button"
      class="btn btn-danger"
    >
      Borrar
    </button>
  </li>
</div>
`;
